crates/sui-adapter/src/programmable_transactions/linkage_view.rs

```diff
@@ -0,0 +1,258 @@
+        // necessary to serve "defining module" requests for unpublished packages, but will also
+        // speed up other requests.
+        for TypeOrigin {
+            module_name,
+            struct_name,
+            package: defining_id,
+        } in context.type_origin_table()
+        {
+            let Ok(module_name) = Identifier::from_str(module_name) else {
+                return Err(ExecutionError::invariant_violation(format!(
+                    "Module name isn't an identifier: {module_name}"
+                )));
+            };
+
+            let Ok(struct_name) = Identifier::from_str(struct_name) else {
+                return Err(ExecutionError::invariant_violation(format!(
+                    "Struct name isn't an identifier: {struct_name}"
+                )));
+            };
+
+            let runtime_id = ModuleId::new(runtime_id, module_name);
+            self.add_type_origin(runtime_id, struct_name, *defining_id)?;
+        }
+
+        Ok(())
+    }
+
+    pub fn storage(&self) -> &'state S {
+        self.state_view
+    }
+
+    fn get_cached_type_origin(
+        &self,
+        runtime_id: &ModuleId,
+        struct_: &IdentStr,
+    ) -> Option<AccountAddress> {
+        self.type_origin_cache
+            .borrow()
+            .get(runtime_id)?
+            .get(struct_)
+            .cloned()
+    }
+
+    fn add_type_origin(
+        &self,
+        runtime_id: ModuleId,
+        struct_: Identifier,
+        defining_id: ObjectID,
+    ) -> Result<(), ExecutionError> {
+        let mut cache = self.type_origin_cache.borrow_mut();
+        let module_cache = cache.entry(runtime_id.clone()).or_default();
+
+        match module_cache.entry(struct_) {
+            Entry::Vacant(entry) => {
+                entry.insert(*defining_id);
+            }
+
+            Entry::Occupied(entry) => {
+                if entry.get() != &*defining_id {
+                    return Err(ExecutionError::invariant_violation(format!(
+                        "Conflicting defining ID for {}::{}: {} and {}",
+                        runtime_id,
+                        entry.key(),
+                        defining_id,
+                        entry.get(),
+                    )));
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl From<&MovePackage> for LinkageInfo {
+    fn from(package: &MovePackage) -> Self {
+        Self {
+            link_context: package.id().into(),
+            original_package_id: package.original_package_id().into(),
+            linkage_table: package.linkage_table().clone(),
+        }
+    }
+}
+
+impl<'state, S: StorageView> LinkageResolver for LinkageView<'state, S> {
+    type Error = SuiError;
+
+    fn link_context(&self) -> AccountAddress {
+        if let Some(LinkageInfo { link_context, .. }) = &self.linkage_info {
+            *link_context
+        } else {
+            AccountAddress::ZERO
```

Comment:
Is this an error condition? In other words, should it be ever returned, what are the consequences for the caller (crash? somehow handled gracefully in the linker?).

Reply:
This is used to implement backwards compatibility -- if package upgrades are disabled, we should be running in a mode where the link context never changes.